### PR TITLE
[Sparse] SIGN Example

### PR DESCRIPTION
**Pull request by mufeili**
## Description

- SIGN example using `mock_sparse`, with accuracy around 0.74~0.78
- Fix two minor issues within `mock_sparse` that occur with gpu usage

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
1. I think this example will use our C++ sparse library in the end. So do we need to add a TODO to indicate we need to test this example after we have the C++ sparse library ready?
2. A README file would help to understand this example.

File: examples/sparse/sign.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        return (Y_pred == Y).float().mean().item()
+
+
+dev = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+data = CoraGraphDataset()
+g = data[0].to(dev)
+src, dst = g.edges()
+N = g.num_nodes()
+A = create_from_coo(dst, src, shape=(N, N))
```
Comment (frozenbugs) [inline examples/sparse/sign.py:46]: In the paper W was used to represent adjacency matrix.
I am not sure about whether the community convention is to use A for adjacency matrix.
Here are 2 suggestions I would have.
1. Use the annotation in the paper directly in our code, and add a comment say, this is not dgl convention, but we are follow the paper.
2. Use the convention, but add a comment to clarify the annotation in the paper was W.

File: examples/sparse/sign.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import torch
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply examples/sparse/sign.py:1]: Also mention in this example, we are implementing a 2 hop sign.

Comment (frozenbugs) [inline examples/sparse/sign.py:1]: https://quip-amazon.com/OWYyAbVeGhBz/DGL-Examples

For the path, let's use core/sparse/sign(_pytorch?).py

File: examples/sparse/sign.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        super().__init__()
+
+        num_feats = num_hops + 1
+        self.linear = nn.ModuleList(
+            [nn.Linear(in_size, hidden_size) for _ in range(num_feats)]
+        )
+        self.pred = nn.Linear(hidden_size * num_feats, num_classes)
+
+    def forward(self, X):
+        X = torch.cat([self.linear[i](X[i]) for i in range(len(X))], dim=1)
```
Comment (frozenbugs) [inline examples/sparse/sign.py:28]: Nesting this together made it hard to read (for me).
Not sure we can unnest the code and add comment to guide readers to understand the code better.

**Comment by mufeili** [comment]
> 1. I think this example will use our C++ sparse library in the end. So do we need to add a TODO to indicate we need to test this example after we have the C++ sparse library ready?

We should. Any suggestion on how to do that?

> 2. A README file would help to understand this example.

Added

File: examples/sparse/sign.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def evaluate(X, Y, model):
+    model.eval()
+    with torch.no_grad():
+        logits = model(X)
+        _, Y_pred = torch.max(logits, dim=1)
+        return (Y_pred == Y).float().mean().item()
+
+
```
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:39]: If users need to use a different device, they can simply edit L40.

I aim to develop an extremely simple script that people can quickly understand what happens without reading any comments.

File: examples/sparse/sign.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+data = CoraGraphDataset()
+g = data[0].to(dev)
+src, dst = g.edges()
+N = g.num_nodes()
+A = create_from_coo(dst, src, shape=(N, N))
+
+I = identity(A.shape, device=dev)
+A_hat = A + I
+D_hat = diag(A.sum(dim=1)) ** -0.5
```
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py]: Nice catch, corrected

File: examples/sparse/sign.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+A_hat = D_hat @ A_hat @ D_hat
```
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:51]: I think this is fine. By doing so people know that `A + I` is no longer needed.

File: examples/sparse/sign.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:1]: Linked the paper in the README file that @czkkkkkk suggested.

[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:1]: The point of declaring a variable `num_hops = 2` is to tell users that we are using 2 hops.

File: examples/sparse/sign.py (hunk 1 above)
Comment (mufeili) [reply examples/sparse/sign.py:46]: It has been the standard to denote the adjacency matrix by A in graph theory, see https://en.wikipedia.org/wiki/Adjacency_matrix. The SIGN paper actually also used it in various spaces like equation (4) in https://arxiv.org/pdf/2004.11198.pdf.

File: examples/sparse/sign.py (hunk 2 above)
Comment (mufeili) [reply examples/sparse/sign.py:1]: Let's discuss it in the meeting next Tuesday.

**Comment by czkkkkkk** [comment]
> > 1. I think this example will use our C++ sparse library in the end. So do we need to add a TODO to indicate we need to test this example after we have the C++ sparse library ready?
> 
> We should. Any suggestion on how to do that?
> 
I simply suggest to add a TODO here.

File: examples/sparse/sign.py (hunk 1 above)
Comment (frozenbugs) [reply examples/sparse/sign.py:46]: In the paper A was defined as A = D−1/2WD−1/2. My understand is, A in equation (4) also means this.

My feeling is, if we also confuses with the annotation, then it would happen to many other readers.
If using A is standard, then add a comment to point out instead of A, W was used in the paper for the adjacency matrix should be enough,

File: examples/sparse/sign.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:39]: The page you referenced is a tutorial for new comers. The example is more for demonstrating the use of sparse matrix library. I prefer not mixing multiple objectives.

File: examples/sparse/sign.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply examples/sparse/sign.py:1]: Newbies should refer to a set of dedicated tutorials.

**Comment by mufeili** [comment]
> > > 1. I think this example will use our C++ sparse library in the end. So do we need to add a TODO to indicate we need to test this example after we have the C++ sparse library ready?
> > 
> > 
> > We should. Any suggestion on how to do that?
> 
> I simply suggest to add a TODO here.

Added a TODO in #4860

File: examples/sparse/sign.py (hunk 1 above)
Comment (mufeili) [reply examples/sparse/sign.py:46]: It's often to see inconsistencies in notations across papers. For example, the symmetric adjacency matrix normalization was probably first known by many people with the [GCN paper](https://arxiv.org/abs/1609.02907), which is way more impactful than the SIGN paper. I follow its notations in naming variables.

**Review by czkkkkkk — APPROVED** [review]

File: examples/sparse/sign.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
+"""
+Simplified version of SIGN, where we only consider symmetrically normalized
+adjacency matrix A and its powers for diffusion operators
```
Comment (frozenbugs) [inline examples/sparse/sign.py:3]: nit: end with dot.

File: examples/sparse/sign.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        X = A @ X
+        X_sign.append(X)
+    return X_sign
+
+
+class Model(nn.Module):
+    def __init__(self, in_size, num_hops, num_classes, hidden_size=256):
+        super().__init__()
+
+        # extra one for the original node features
```
Comment (frozenbugs) [inline examples/sparse/sign.py:26]: Extra one for the original node features.

File: examples/sparse/sign.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        num_feats = num_hops + 1
```
Comment (frozenbugs) [inline examples/sparse/sign.py:27]: num_feats is a little bit mis-leading, do we have a better name?

File: examples/sparse/sign.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    X_sign_train.append(feat[train_mask])
+    X_sign_test.append(feat[test_mask])
+Y_train = Y[train_mask]
+Y_test = Y[test_mask]
+
+model = Model(X.shape[1], num_hops, data.num_classes).to(dev)
+loss_func = nn.CrossEntropyLoss()
+optimizer = Adam(model.parameters(), lr=3e-3)
+
+model.train()
```
Comment (frozenbugs) [inline examples/sparse/sign.py:82]: What does this do?

File: examples/sparse/sign.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+for epoch in range(10):
+    pred = model(X_sign_train)
+    loss = loss_func(pred, Y_train)
+    optimizer.zero_grad()
+    loss.backward()
+    optimizer.step()
+    train_acc = evaluate(X_sign_train, Y_train, model)
```
Comment (frozenbugs) [inline examples/sparse/sign.py:89]: For all example output, can we use a standard output format?

File: examples/sparse/sign.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    print(f"epoch {epoch} | train acc {train_acc}")
```
Comment (frozenbugs) [inline examples/sparse/sign.py:90]: I tried the code, it is a little bit overfit, is it a bug?


epoch 0 | train acc 0.27142858505249023
epoch 1 | train acc 1.0
epoch 2 | train acc 0.9428571462631226
epoch 3 | train acc 0.9285714030265808
epoch 4 | train acc 0.9642857313156128
epoch 5 | train acc 0.9857142567634583
epoch 6 | train acc 0.9928571581840515
epoch 7 | train acc 0.9928571581840515
epoch 8 | train acc 0.9928571581840515
epoch 9 | train acc 0.9928571581840515
test acc 0.7490000128746033

File: examples/sparse/sign.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
+"""
```
Comment (frozenbugs) [inline examples/sparse/sign.py:1]: Let's adopt the similar style (model definition + train method) as we established in https://github.com/dmlc/dgl/tree/master/tutorials/blitz.

File: examples/sparse/sign.py (hunk 7 above)
Comment (mufeili) [reply examples/sparse/sign.py:3]: This is not a valid sentence.

File: examples/sparse/sign.py (hunk 10 above)
Comment (mufeili) [reply examples/sparse/sign.py:82]: This is mostly for handling the different behaviors of batch normalization and dropout between training and evaluation. It's not necessary for this example though. For a reference, see https://pytorch.org/tutorials/beginner/saving_loading_models.html?highlight=eval

File: examples/sparse/sign.py (hunk 12 above)
Comment (mufeili) [reply examples/sparse/sign.py:90]: This is often the case for Cora. There are only 140 training nodes in Cora after all.

**Comment by mufeili** [comment]
Closed via #4908
